app/models/manageiq/providers/hawkular/middleware_manager.rb

```diff
@@ -262,6 +262,14 @@ def add_middleware_deployment(ems_ref, hash)
           :binary_content        => hash[:file]["file"].read,
           :resource_path         => ems_ref.to_s
         }
+        unless hash[:file]['server_groups'].nil?
+          # in case of deploying into server group the resource path should point to the domain controller
+          deployment_data[:server_groups] = hash[:file]['server_groups']
+          server_group_path_hash = ::Hawkular::Inventory::CanonicalPath.parse(deployment_data[:resource_path]).to_h
+          server_group_path_hash[:resource_ids].slice!(1..-1)
```

Comment:
the order is guaranteed and it has the semantics that the very first is the root in the parent-child hierarchy.

so for instance `resource_ids = [1,2,3]` means 
`1 --contains--> 2 --contains--> 3`

The Hawkular inventory is a graph database that has its resources organized as a graph. For the `contains` edge, it's guaranteed, that it's actually a tree. We use the so called canonical paths to address all the entities in the inventory.

In this particular case what I was trying to do is obtaining the domain controller canonical path from the server group c. path. For the WildFly running in the domain mode, it's easy, because the root of the graph is guaranteed to be a host controller. That's why we just pick the first id from the `resource_ids` obtained from the server group.

Perhaps the future improvement would be to add the more readable methods to the `CanonicalPath` class, like `.parent(hops)`, `.root()`